zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,309 @@
+//! The compress circuit implementation.
+
+use crate::evm_circuit::util::from_bytes;
+use crate::util::{Expr, SubCircuitConfig};
+use bit_vec::BitVec;
+use eth_types::Field;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter, Region, Value},
+    plonk::{Advice, Column, ConstraintSystem, Error, Selector, TableColumn},
+    poly::Rotation,
+};
+use lazy_static::lazy_static;
+use std::marker::PhantomData;
+
+lazy_static! {
+    // TODO: maybe fixed huffman table will get from witness
+    static ref FIXED_HUFFMAN_TABLE: [BitVec; 256] = [(); 256].map(|_| BitVec::new());
+}
+
+mod combine {
+    const LEFT_SHIFT_8BITS: u64 = 256;
+
+    use crate::util::Expr;
+    use eth_types::Field;
+    use halo2_proofs::plonk::Expression;
+    pub(crate) fn expr<F: Field>(parts: [Expression<F>; 32]) -> Expression<F> {
+        parts
+            .into_iter()
+            .fold(0.expr(), |acc, part| acc * F::from(LEFT_SHIFT_8BITS) + part)
+    }
+}
+
+mod compress {
+    use super::FIXED_HUFFMAN_TABLE;
+    use crate::evm_circuit::util::from_bytes;
+    use eth_types::Field;
+
+    pub(crate) fn value<F: Field>(src: F) -> [F; 32] {
+        src.to_repr().map(|item| {
+            let code = FIXED_HUFFMAN_TABLE[item as usize].to_bytes();
+            from_bytes::value::<F>(&code)
+        })
+    }
+}
+
+mod rlc {
+    use crate::util::Expr;
+    use eth_types::Field;
+    use halo2_proofs::plonk::Expression;
+
+    pub(crate) fn expr<F: Field>(
+        parts: [Expression<F>; 32],
+        randomness: Expression<F>,
+    ) -> Expression<F> {
+        parts
+            .into_iter()
+            .rev()
+            .fold(0.expr(), |acc, part| acc * randomness.clone() + part)
+    }
+
+    pub(crate) fn value<F: Field>(src: [F; 32], randomness: F) -> F {
+        src.into_iter()
+            .rev()
+            .fold(F::zero(), |acc, part| acc * randomness + part)
+    }
+}
+
+// layout
+// ```
+// | q_enable | q_end | rand_rpi | rpi0 | rpi1 | ... | rpi31 | cpi0 | cpi1 | ... | cpi31 |  cpi_rlc_acc | huffman_value | huffman_code |
+// ------------------------------------------------------------------------------------------------------------------------------------
+// | 1        | 0     | rand     |rpi[0]|rpi[1]| ... |rpc[31]|cpi[0]|cpi[1]| ... |cpi[31]| ........... | ............. | ............ |
+// | 1        | 1     | rand     |rpi[0]|rpi[1]| ... |rpc[31]|cpi[0]|cpi[1]| ... |cpi[31]| ........... | ............. | ............ |
+// ```
+
+/// Config for compress circuit
+#[derive(Clone, Debug)]
+pub struct CompressCircuitConfig<F: Field> {
+    q_enable: Selector,
+    q_end: Selector,
+    raw_public_inputs: [Column<Advice>; 32],
+    compressed_public_inputs: [Column<Advice>; 32],
```

Comment:
Was there a specific reason you chose to store 32 bytes/row? In any case, like I said on the call, decoding one Huffman encoded word/row makes more sense to me and should be sufficient. With 32 bytes per row in the worst case we'd have to do a lookup/bit to decode the data which would mean 32*8=256 lookups!

Reply:
I think we need to split origin_rpi's row into 32 rows, u256 -> [u8;32]
Using 32bytes/row is easy to constraint between the `origin raw public inputs` and `split raw public inputs`
I have no sense to constraint between them if we use the u8/row

| orpi | rpi|
| ----- |---------- |
| rpi[0] | rpi[0][0] |
| rpi[1] | rpi[0][1] |
|           | rpi[0][2] |
|           | ......         |
|           | rpi[1][31] |

orpi: origin raw public inputs
rpi: split raw public inputs(into bytes stream

Reply:
> With 32 bytes per row in the worst case we'd have to do a lookup/bit to decode the data which would mean 32*8=256 lookups!

Now, we just encode the raw_public_inputs, and compare with compressed_public_inputs from l1 constract.
So, we lookup value(byte) to code

Reply:
I would say don't worry too much about getting all the 32 bytes on a single row again for now (this will still change when the circuit switches to keccak which is also byte based). But the mechanism it the same as you'll need to collect the (un)compressed data stream to check if it matches the original data stream. Whether you see it as encoding (variable number of compressed bits/row) or decoding (variable number of processed bits/row) you'll have to somehow collect those bits/bytes over multiple rows. So you just use the rlc trick with some extra logic.

So the easy example is like this, where on each row simply a single byte is added to the rlc:

| data | acc |
| -----  |---------- |
| byte[0] | acc = byte[0] |
| byte[1] | acc = (acc, rot::prev()) * r + byte[1] |
| byte[2] | acc = (acc, rot::prev()) * r + byte[2] |
| .... | .... |
 .... | data_rlc |

At the end you get the `data_rlc` which it the encoding of all bytes. However, for the compression part or even if the circuit has to consume multiple bytes per row (like needing 32 bytes on some rows) the logic just gets a bit more complicated. On some rows you add 32 bytes to `acc`, on other row you may not add any bytes to it. For the compression/decompression circuit you may even have to do this on the bit level where you only want to add a couple of bits to the rlc (depends on the implementation because you can also think of ways to keep the rlc accumulation on the byte level). At the end though you'll have to end up with a value that is equal to `data_rlc`, so that's simply a check you add at the bottom of the circuit.

> Now, we just encode the raw_public_inputs, and compare with compressed_public_inputs from l1 constract.
So, we lookup code by value(byte)

Ah yes that would indeed mean only up to 32 lookups. That' still a lot though (no way around this many lookups for some circuits but here seems pretty inefficient to require that).

Reply:
I'm sorry, I didn't make it clear, I want to say that how can I constraint the `txlist` between the `tx_table`'s format and the `raw_public_inputs`(byte/row)'s format

Reply:
1. The tx field in `tx_table` maybe has been rlc encoded, but the `raw_public_inputs` that will be compressed into txlist doesn't have rlc encoded fields
2. The same tx field of txlist in `tx_table` has been put which is easy for using rotation